plotly/tools.py

```diff
@@ -2871,3 +2891,149 @@ def get_candle_decrease(self):
         return (decrease_x, decrease_close, decrease_dif,
                 stick_decrease_y, stick_decrease_x)
 
+class _Dendrogram(FigureFactory):
+    ''' Returns a Dendrogram figure object
+        Example usage:
+        D = Dendrogram( Z )
+        fig = { 'data':D.data, 'layout':D.layout }
+        py.iplot( fig, filename='Dendro', validate=False )'''
+    
+    def __init__(self, X, orientation='bottom', labels=None, colorscale=None, \
+                 width=700, height=700, xaxis='xaxis', yaxis='yaxis' ):
+        ''' Draw a 2d dendrogram tree 
+            X: Heatmap matrix as array of arrays
+            orientation: 'top', 'right', 'bottom', or 'left'
+            labels: List of axis category labels
+            colorscale: Optional colorscale for dendgrogram tree clusters
+            Returns a dendrogram Plotly figure object '''
+    
+        self.orientation = orientation
+        self.labels = labels
+        self.xaxis = xaxis
+        self.yaxis = yaxis
+        self.data = []
+        self.leaves = []
+        self.sign = { self.xaxis:1, self.yaxis:1 }        
+        self.layout = { self.xaxis:{}, self.yaxis:{} }
+        
+        self.sign[self.xaxis] = 1 if self.orientation in ['left','bottom'] else -1
+        self.sign[self.yaxis] = 1 if self.orientation in ['right','bottom'] else -1        
+        
+        dd_traces, xvals, yvals, ordered_labels, leaves = self.get_dendrogram_traces( X, colorscale )
+        
+        self.labels = ordered_labels
+        self.leaves = leaves
+        yvals_flat = yvals.flatten()
+        xvals_flat = xvals.flatten()
+
+        self.zero_vals = []
+
+        for i in range(len(yvals_flat)):
+            if yvals_flat[i] == 0.0 and xvals_flat[i] not in self.zero_vals:
+                self.zero_vals.append( xvals_flat[i] )
+
+        self.zero_vals.sort() 
+                
+        self.layout = self.set_figure_layout( width, height )
```

Comment:
Style note, pep:8ball: on the whitespace around these args: https://www.python.org/dev/peps/pep-0008/#whitespace-in-expressions-and-statements

Rather:

`self.layout = self.set_figure_layout(width, height)`